[KTextbox] Add slots to append elements before or after input

Abhishek-Punhani opened the pull request:
<!-- Please remove any unused sections -->

## Description
<!-- What does this PR do? Briefly describe in 1-2 sentences* -->
- Added four new named slots (`outerBefore`, `outerAfter`, `innerBefore`, `innerAfter`) to `KTextbox/UiTextbox` for flexible content injection before/after the input, both inside and outside the input area.
- Fixed fallback rendering for icon props so that they display correctly when no slot is provided.

#### Issue addressed
<!-- Only necessary if applicable -->

Fixes #1113 

### Before/after screenshots
<!-- Insert images here if applicable -->
<img width="521" height="127" alt="Screenshot from 2025-09-12 14-36-22" src="https://github.com/user-attachments/assets/4c37e68a-345e-4347-a4ba-6f16965c61b6" />

- https://github.com/learningequality/studio/pull/5303
<img width="679" height="84" alt="Screenshot from 2025-09-12 14-41-30" src="https://github.com/user-attachments/assets/339649fd-808c-450f-a981-1daaba60fcc0" />


## Changelog
<!-- [DO NOT REMOVE-USED BY GH ACTION] CHANGELOG START -->

<!--
  - Fill in the changelog item(s) below. If there are more groups of closely
    related changes, prepare more changelog items for each one of them.
    At a minimum, always separate non-breaking changes from breaking changes.
  - This needs to be pasted to CHANGELOG.md before merging a PR.
  - See changelog guidelines https://www.notion.so/learningequality/DRAFT-Changelog-Guidelines-106b6ebbdeda4ba5b3b3e7c490c5a4fe and existing
    items in CHANGELOG.md as examples
 -->

  - **Description:** Add `outerBefore`, `outerAfter`, `innerBefore`, `innerAfter` slots to `KTextbox` to append elements before or after input
  - **Products impact:** KTextbox
  - **Addresses:** #1113 
  - **Components:** : KTextbox, UiTextbox
  - **Breaking:** No
  - **Impacts a11y:** No
  - **Guidance:** -

<!-- [DO NOT REMOVE-USED BY GH ACTION] CHANGELOG END -->

## Steps to test

1. Use KTextbox with the new slots in a consumer app or storybook.
2. Pass content to each slot and verify it renders in the correct location.
3. Test with and without label/icon props and slots to confirm fallback logic.


### At a high level, how did you implement this?
- Updated UiTextbox template to add four named slots and improved conditional rendering for icon.
- Updated KTextbox to forward all relevant slots and props.

### Does this introduce any tech-debt items?
We will need to update the KDS website KtextBox docs

## Testing checklist
<!-- Complete the checklist before submitting a PR; delete anything that doesn't apply -->

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical and brittle code paths are covered by unit tests
- [ ] The change is described in the changelog section above

## Reviewer guidance
<!-- Delete anything that doesn't apply so your reviewer knows what to check for -->

- [ ] Is the code clean and well-commented?
- [ ] Are there tests for this change?
- [ ] Are all UI components LTR and RTL compliant (if applicable)?
- [ ] _Add other things to check for here_

## Comments

Abhishek-Punhani committed:
Feat : Add slots to append elements before or after input

MisRob commented:
Thank you @Abhishek-Punhani. I've just returned from a vacation and need some time to catch up on everything. Then I will review.

Abhishek-Punhani commented:
> Thank you @Abhishek-Punhani. I've just returned from a vacation and need some time to catch up on everything. Then I will review.

Sure. Let me know if any changes are required

MisRob commented:
@Abhishek-Punhani would you be able to add visual tests and documentation examples as mentioned in the issue's requirements? It would help me to test everything.

MisRob commented on lib/KTextbox/index.vue:
```diff
@@ -53,6 +70,13 @@
         required: true,
       },
+      /**
+       * Icon for the text field (uses UiIcon names)
```
I don't quite understand why is this prop and also `#icon` slot is needed for the task that this PR attempts to solve?

I think the intention was to just provide `KIcon` via one of the four new slots. Was there a use-case where that's not suitable?

Abhishek-Punhani replied:
@MisRob, I noticed that in `lib/keen/UiTextbox.vue`, it was expecting both a prop and a slot, but it wasn't passed in index.vue, so I just added it. Let me know if we want to be just a prop or a slot.

MisRob replied:
I will think about it, thanks for info

MisRob replied:
I removed `UiIcon` logic completely - we don't use it and generally, we're trying to cleanup unused code from `Ui..` components.

MisRob reviewed (commented):
Thank you @Abhishek-Punhani, overall this direction looks good to me. I left few initial notes. I appreciated the screenshot, even though to be able to merge this feature, we would need to have visual tests and documentation updated anyway. I will do one more detailed review round as soon as those are in place - it will help me to confirm all works as expected.

Abhishek-Punhani commented:
> Thank you @Abhishek-Punhani, overall this direction looks good to me. I left few initial notes. I appreciated the screenshot, even though to be able to merge this feature, we would need to have visual tests and documentation updated anyway. I will do one more detailed review round as soon as those are in place - it will help me to confirm all works as expected.

@MisRob, I will also add visual tests for this feature. Could you please provide some guidance on updating documentation? I just need to add docsExample demonstrating all slots, right?

MisRob commented:
Thank you @Abhishek-Punhani.

>  I just need to add docsExample demonstrating all slots, right?

Exactly. And actually - example files are shared between visual tests and documentation :) It's explained in [the visual testing guidance](https://github.com/learningequality/kolibri-design-system/blob/develop/dev_docs/07_visual_testing_guide.md) that we've recently updated - so I recommend to follow step by step. Then you can just import those very same examples to the documentation. No need to write much - I will fine-tune documentation details before we merge if needed.

Abhishek-Punhani commented:
@MisRob, I have added `DocExamples`, but I am getting errors with Percy in localhost 
``` bash
FAIL lib/KTextbox/__tests__/visual.spec.js
  ● Test suite failed to run

    The error below may be caused by using the wrong test environment, see https://jestjs.io/docs/configuration#testenvironment-string.
    Consider using the "jsdom" test environment.

    ReferenceError: document is not defined

      68 |     };
      69 |
    > 70 |     if (document.readyState === 'loading') {
         |     ^
      71 |       document.addEventListener('DOMContentLoaded', onDomReady);
      72 |     } else {
      73 |       onDomReady();

      at KThemePlugin (lib/KThemePlugin.js:70:5)
      at Function.Vue.use (node_modules/vue/dist/vue.runtime.common.dev.js:5731:20)
      at Object.use (jest.conf/setup.js:41:5)
      at Object.require (jest.conf/visual.setup.js:1:1)

Test Suites: 1 failed, 1 total

```

Abhishek-Punhani committed:
Adding DocExamples for slots added to KTextBox

Signed-off-by: Abhishek-Punhani <punhani.manavabhi@gmail.com>

MisRob commented:
@Abhishek-Punhani As for the error, I haven't yet noticed anything in code that should be causing it. I will try on my end. Thanks for adding the examples.

MisRob committed:
Finalize documentation

MisRob committed:
Remove unused UiIcon related logic

It is not publicly exposed from KTextbox,
neither used in consumers.

To place icon before or after the input,
the new outer/inner after/before slots can be used.

MisRob committed:
Make styling from consumers easier

by not being opinionated which styles
KDS applies to slots.

Default styles are helpful in some cases,
however overriding UiTextbox from consumers is
difficult and we don't yet have sufficient amount
of use-cases to see what's common patterns.

Also don't render slot divs when slots are not
being used.

MisRob committed:
Simplify internal styles

and markup structure by removing
unnecessary wrappers. Same effect can be
achieved without them.

MisRob committed:
Fix incorrect position of label

When inner slots were present, the label
was centered, but it should be on the left
side. Centered position was likely use to fix
overlapping of the label with the content
of the left inner slot in an empty input.
This is now resolved by ensuring that the label
is positioned in the upper area of the input
(=not applying transform style) when an inner slot
is present.

MisRob committed:
Stretch slot area vertically

to make positioning of its content
from consumers easier.

MisRob committed:
Add slots documentation

MisRob committed:
Remove obsolete file

Example was moved to another file.

MisRob committed:
Return style that I removed by mistake

MisRob commented:
Hi @Abhishek-Punhani - I reviewed everything and very nice work. Thank you.

As mentioned, I built on top of the documentation and tweaked it to represent Kolibri and Studio use-cases.

I had two suggestions around how to (1) simplify inner styling, (2) make slot content's styling easier from consumers. However I weren't sure if they will work, so as I was finalizing documentation examples, I played around with some approaches and made those updates. I left notes on each change in commit messages so you still can receive some feedback.

There was only one regression when the label was centered when using one of the inner slots. It seemed to be an intentional change to ensure no overlap with the label. I resolved it differently so that we don't need to center the label. Again see the commit message for details.

Since we both did some significant changes of `UiTextbox` internals, some of them having a potential to take effect even in scenarios when slots are not being used, before we merge I'd like to see that there's no regressions in Kolibri. I will soon open a Kolibri QA PR. After QA is done, we can merge this.

MisRob commented:
As for the Percy tests, they run just fine on the cloud

<img width="200" height="auto" alt="Screenshot from 2025-10-13 14-17-01" src="https://github.com/user-attachments/assets/501185bd-d226-4569-a5c4-52d35b1f2e47" />

MisRob commented:
@Abhishek-Punhani actually I've just noticed one thing that I think was present before my changes too.

If you'd like to look into why it's happening, that'd be great. Don't forget to pull my updates first though :)

<img width="500" height="auto" alt="bug-input" src="https://github.com/user-attachments/assets/61038828-c77d-40db-bd22-2e7e267bea9f" />

And feel free to review the changes I pushed and let me know if you had any questions or noticed anything that I may have missed.

Abhishek-Punhani commented:
@MisRob , Thank you so much for the thorough review and for taking the time to build on top of the documentation.
I’ve reproduced the changes you made to the inner styling and slot content on my local setup — everything looks good except for the regression you mentioned. I will look into it and update my PR accordingly as soon as possible.

I really appreciate your guidance and feedback on this.

Abhishek-Punhani committed:
Removing unnecessary border

Signed-off-by: Abhishek-Punhani <punhani.manavabhi@gmail.com>

Abhishek-Punhani commented:
@MisRob , I noticed that on disabled and readonly conditions, a border is added to both `.ui-textbox-input`  and  ` .ui-textbox-label`, which is causing this double border. I think we can remove one border in this specific condition of read-only and disabled, without affecting other CSS cases.
Please let me know your thoughts on this.

MisRob commented:
Thanks @Abhishek-Punhani, yes I think the way you fixed the border makes sense. Before we merge, we will do [regression testing in Kolibri](https://github.com/learningequality/kolibri/pull/13839).

MisRob reviewed (approved):
Good news @Abhishek-Punhani, we haven't observed any regressions.

Very nice work here, thanks a lot - this will not only make possible finishing your related Studio PR, but unblock many other tasks.

Also thanks for starting the documentation and visual tests. I know I tweaked those quite a bit, but it's much easier to do when there's some basis. This is what I do normally as I am trying to oversee the documentation quite carefully as a whole.

We will soon release a new version of KDS and install it in Studio. I will let you know when that's done.
